### ISSUE TITLE
0.4 roadmap

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#11536]
Done:

- [x] refactor abstract array type (#10525)
- [x] change colon lowering (#10331)
- [x] serializing with cycles (#10170)
- [x] bare doc strings? (#8965)

Remaining:

- [x] package precompilation (#8745)
- [x] reserve field default syntax (#10146)
- [ ] ~~make [a,b] non-concatenating (#7128)~~

Potential:

- [ ] reserve triangular dispatch syntax?
- [ ] convert default values?

No consensus reached:

- [ ] ~~ordered dicts? (#10116)~~
- [ ] ~~zero initialization? (#9147)~~

Deferred:

- [ ] ~~defer method ambiguities to runtime?~~
- [ ] ~~change constructor syntax? (cc @JeffBezanson)~~

**Comment by tkelman** [comment]
#10170 was already merged. I think #10331 is a part of #10525 but @mbauman can correct me if I'm wrong.

**Comment by mbauman** [comment]
>  I think #10331 is a part of #10525

Not quite - #10525 allows indexing with `Colon()`, but it doesn't change the parser lowering of `[:]`.  That's what #10331 provides.

**Comment by JeffBezanson** [comment]
I'm going to move all strictly-bug and -performance issues from the 0.4 milestone to 0.4.1.

**Comment by tkelman** [comment]
Better to keep the distinction between 0.4.x and 0.4.1, otherwise that's going to get really annoying when it's time to actually release 0.4.1 as a bugfix release.

**Comment by JeffBezanson** [comment]
Yes, I think I meant 0.4.x.

**Comment by sbromberger** [comment]
I really hope we can keep #9147 in 0.4-release... this one just bit me again.

**Comment by jakebolewski** [comment]
Why is #9147 on the release roadmap?  Of all the changes to make it does not really seem to be the most pressing one, and it can always be added in the future as a non-breaking change.

**Comment by JeffBezanson** [comment]
I'd really like to have it, but you're right that it's non-breaking and technically doesn't need to be release blocking.

**Comment by sbromberger** [comment]
It's not critical to the release but it's a security-significant change that fixes an issue that's caused large amounts of grief to various packages.

In the event we don't get this into the release (which would be a tremendous shame) would you be amenable to a documentation PR that states, effectively and emphatically, **DO NOT TRUST UNINITIALIZED VARIABLES**?

**Comment by carnaval** [comment]
I don't see why zero fixes it though. At least garbage is obvious most of the time (we could actually have a guaranteed garbage mode as a debug thing to spot usage of unintialized stuff). For example I use a small integer field as a flag for some things and zero has a very special meaning. I find it dangerous to assume that zero is always a safe choice.

**Comment by sbromberger** [comment]
@carnaval I really think we should keep the discussion of the merits of the approach in the actual issue thread itself so we don't get two distinct threads.

**Comment by JeffBezanson** [comment]
Yes, please see the discussion in that issue, where many such points have been covered.

**Comment by ScottPJones** [comment]
What about Unicode bug fixes? :grinning:

**Comment by quinnj** [comment]
@ScottPJones, bug fixes are great for bug fix releases :) It's mainly breaking, larger behavior changes, syntax changes, etc. that are important to get in before an official 0.4 release so that those features are available for the entire 0.4.x release cycle where pretty much *only* bug fixes/performance improvements are allowed.

**Comment by prcastro** [comment]
What does "bare doc strings" mean?

**Comment by StefanKarpinski** [comment]
"bare doc strings" means being able to use plain strings in void context to provide doc strings instead of needing to use `@doc` macros.

**Comment by ScottPJones** [comment]
@quinnj But if the code is already ready, and many people have already said it should go in?  Bug fixes help move the language forward also, they make the language more reliable.

**Comment by tkelman** [comment]
This issue is about which major breaking issues are going to determine scope completion for moving from 0.4-dev to 0.4-pre (feature freeze). Non-breaking bug fixes can continue at any time, though we will be especially conservative about only committing fixes for release-blocker bugs as we get closer to tagging the release (after branching for release-0.4, for example).

**Comment by timholy** [comment]
@ScottPJones, it's likely that there will be weeks/months between feature freeze and release, and the enthusiasm for bugfixes is especially high during that period. Let's keep this issue focused on breaking changes.

**Comment by ScottPJones** [comment]
@timholy Since certain people have said that they think my bug fixes might be breaking, wouldn't that mean that those changes definitely *should* go in earlier, rather than later?

**Comment by ViralBShah** [comment]
Bare doc strings are essential for package help, and I really would like to see those happen as part of the 0.4 release.

**Comment by ViralBShah** [comment]
There are 20 items in all marked as 0.4.0. Should we trim that label to "must do" items from this issue, and mark any important bug fixes slated for 0.4.0 to 0.4.x? That way, there is a clear line of sight to the 0.4 release.

**Comment by ScottPJones** [comment]
:+1: to bare doc strings

**Comment by prcastro** [comment]
:+1: to bare doc strings here also

**Comment by tkelman** [comment]
> Should we trim that label to "must do" items from this issue, and mark any important bug fixes slated for 0.4.0 to 0.4.x?

I think Jeff already did that to a large extent. Anything he left in 0.4.0 was probably for good reason?

**Comment by ViralBShah** [comment]
I guess given that @lindahua is busy, is SparseVectors realistically possible. Perhaps it fits well with the Arraymageddon release anyways.

**Comment by ViralBShah** [comment]
Also, I am not sure we can take vector transposes seriously in 0.4, and Arraymageddon is the right release for it anyways.

**Comment by ViralBShah** [comment]
Range indexing producing subarrays is another thing that should be in the Arraymageddon release.

**Comment by tkelman** [comment]
Probably true on all counts. I'd personally like to see SparseVectors sooner assuming they could be finished in time before feature freeze, but it's not worth holding up a release over.

**Comment by ViralBShah** [comment]
I've moved all those to 0.5.

**Comment by PallHaraldsson** [comment]
@quinnj: "bug fixes are great for bug fix releases :) It's mainly breaking, larger behavior changes, syntax changes, etc. that are important to get in before an official 0.4 release so that those features are available for the entire 0.4.x release cycle"

https://github.com/JuliaLang/julia/issues/9297 this is breaking and a bug and then get in 0.4? I haven't reviewed Scott code yet.. it might be ok to wait for 0.4.1 unlike this one. Is it hard to just disable array access to strings?

**Comment by JeffBezanson** [comment]
On my last 0.4 milestone re-triage I left the sparse and linear algebra issues for others to decide on.

**Comment by malmaud** [comment]
My grandfather's dying wish was that sparse vectors be included in 0.4.

**Comment by mauro3** [comment]
I just read through this issue and picked up these potentially missing items to the list:
```
- [ ] NTuples (#11242)
- [ ] Sparse vectors (#11424)
- [ ] restrict string indexing (#9297)
```
(Note, this doesn't mean I think those should be included, those are just issues other mentioned in this thread.)

**Comment by prcastro** [comment]
-  [ ] Conditional modules (#6195)

:dog:

**Comment by ViralBShah** [comment]
Sparse vectors (#11424) can happen if everyone chips in. Much of the work has been done by @lindahua

**Comment by prcastro** [comment]
#8965 is fixed now

**Comment by timholy** [comment]
Is there anything else that needs to hold up tagging 0.4 candidates?

**Comment by johnmyleswhite** [comment]
Doc conversion merged?

**Comment by StefanKarpinski** [comment]
Yeah, I think that's it. We may also want to let the static compilation thing settle over the weekend.

**Comment by dcjones** [comment]
The cherry on top of 0.4 would be conditional modules in some form. Requires.jl provided a mechanism for this, even if it wasn't perfect, but it seems to be mutually exclusive with precompilation, which means sticking with the ol' "try to load everything" approach, which is unfortunate. Still, I wouldn't want to hold up 0.4 if that needs more work.

**Comment by ViralBShah** [comment]
We should try and get the doc conversion merged as well next week and then call a feature freeze. Assuming that happens, is this roughly what we should expect in the release process?

July 25 - feature freeze
Aug 7 - RC1 (with RCs every week)
Aug 31 - final release (tentatively)

**Comment by ViralBShah** [comment]
It appears that #11943 the only thing that we need to do before feature freeze?

**Comment by vtjnash** [comment]
are we leaving this open so that we can close it with the rc branch-commit?

**Comment by tkelman** [comment]
I think we still want to get the docs at least moved to docstrings in #11943. We can adjust VERSION to `0.4.0-pre` to mark feature freeze, branch for release-0.4, and tag a first release candidate, but I think these could be done in most permutations and not necessarily simultaneously.

**Comment by ViralBShah** [comment]
Now that #11943 is merged, what do we do next here? I feel like we should tag and release the first RC, and sort out the remaining fine tuning of docsystem stuff by RC2 or RC3.

**Comment by tkelman** [comment]
There's a memory leak with an open fix in #12433 that we should fix before doing an RC.

And have you tried REPL help since #11943 was merged? It's pretty much completely broken. #12437

**Comment by ViralBShah** [comment]
I did see the issues. Are you saying that once these two issues are fixed, we should do the RC?

**Comment by tkelman** [comment]
Works for me. I don't think we should do an RC that we wouldn't feel comfortable distributing as if it were a final release, but as soon as the REPL help is back to where it was pre-#11943 I think it should be close enough.

**Comment by amitmurthy** [comment]
FWIW, I have stopped updating docs till the migration is complete. So, maybe a couple of days with the stated intention of doc review/update will be helpful for a good quality RC.

**Comment by tkelman** [comment]
#11943 was the migration. Whether the docs are in RST or Markdown in the docstrings is probably the least consequential issue right now, we should clean things up as best we can right now rather than making 0.4 contingent on reviewing a tens-of-thousands-of-lines reformatting of all docstrings. Are they ending up visible to users in the REPL, and are they making it properly to the manual, are the important questions. Right now the answers are no, and mostly.

**Comment by ViralBShah** [comment]
Ok, there is now #12511 which is ready to merge and it looks like #12437 still needs fixing. What else remains?

**Comment by tkelman** [comment]
Yeah just need a path towards what to do on #12489 and the related doc fixes.

**Comment by MichaelHatherly** [comment]
#12496 would be good to have for the methods that can't be documented at their definition site, but with #12511 that is a relatively small amount now - just those in ``essentials.jl`` and some of the stuff in ``Core``. A bikeshed on the actual syntax for that would be worth having, maybe.

**Comment by MichaelHatherly** [comment]
There's also the "bindings" part of #12088 that would avoid documenting constants by their value rather than name. @one-more-minute would that be simple to split out?

**Comment by tkelman** [comment]
Oh, we also still need to do something about https://github.com/JuliaLang/julia/issues/9079#issuecomment-127274743, freezing due to files with the same names as packages isn't an acceptable thing to do to users. May or may not need to block RC1 for that, but it definitely needs fixing before final.

**Comment by ivarne** [comment]
I think release candidates sounds like a finished product without known blocking issues. When 0.3 was released, [we used the `-dev` mark](https://github.com/JuliaLang/julia/commit/57bdf431b277193de53fe99d0ead210970d17a0) and planed to switch to `-pre` to mark feature freeze (which seems to be about now). The switch would be a signal to package authors that no new breaking changes where pending, so it is time to go through and make sure packages are updated. 

`alpha` and `beta` releases would also be possible, but as our `Version` type orders alphabetically we'll have `alpha < beta < dev` which isn't as nice as `dev < pre < rc`. We also don't have the distinction between internal and external test versions, so I don't think the alpha/beta distinction makes sense. 

I'd like to see `0.4-pre1` and `0.4-pre2` releases, if we want to release versions with known issues.

**Comment by ViralBShah** [comment]
I like the idea of pre releases. We could announce a feature freeze shortly and start with pre releases, switching to RCs then. This will also shorten the RC cycle.

**Comment by ViralBShah** [comment]
I also like the idea of getting a pre-release out right away, so that package authors can start preparing for 0.4 and start plugging into the new doc system.

**Comment by ViralBShah** [comment]
I also wonder if someone should be appointed as a release manager for this release, to make sure all the things happen. @tkelman is the natural choice, but he may not want further distractions in the last phase of his PhD filing.

**Comment by one-more-minute** [comment]
@MichaelHatherly sure, should be simple enough to pull out the bindings commit and redo the macro. I can't compile Julia easily right now so if you're interested go ahead. Would also be good to remove the `objdoc` fallback before an RC.

**Comment by ScottPJones** [comment]
RC should really be just that, a real candidate for release, so no *known* bugs (at least, bugs that people intend to have fixed upon release), so definitely it should be -pre, after feature freeze and branch off of master.

**Comment by MichaelHatherly** [comment]
@one-more-minute I'll have a look at it. Have you had a chance to take a look at #12511? Just rebasing it now.

**Comment by one-more-minute** [comment]
Ah hadn't seen that, will take a look ASAP

**Comment by tkelman** [comment]
How different would a pre-release would be than our usual nightlies. I think now would be a good time to mark a feature freeze by changing `VERSION` from `0.4-dev` to `0.4-pre` though.

If someone wants to help with a simple but slightly tedious issue, the line numbers in `doc/devdocs/init.rst` are all wrong right now which causes `make -C doc linkcheck` to fail, ref https://github.com/JuliaLang/julia/pull/9468#issuecomment-128257377.

**Comment by IainNZ** [comment]
Pre-release in my mind is feature freeze, and is basically an open invitation for the "engaged" community to dive in (e.g. package devs). The RC is the risk-tolerant people, wider set. Not sure if I'm just making things up.

**Comment by ScottPJones** [comment]
@IainNZ "Release candidate", at least in my experience, has a very specific meaning, it should be something that, barring new bugs found in testing, will become the next release.
Whether you call the feature freeze version an alpha, beta, or pre-release version, that's a very different beastie from a real RC.

**Comment by ivarne** [comment]
> How different would a pre-release be than our usual nightlies?

I would guess that a prerelease has a git tag and an announcement, which the nightlies doesn't have. The important part is that we begin to encourage users to check out 0.4, as we will stop pushing breaking changes that require users to change their code (unless they depend upon buggy behavior)

**Comment by ViralBShah** [comment]
Is #12435 also on the critical path here?

**Comment by tkelman** [comment]
Ref https://github.com/JuliaLang/julia/issues/12489#issuecomment-129728362. IMO no.

**Comment by tkelman** [comment]
Anything other than #12742 left before branch and freeze?

**Comment by timholy** [comment]
Would be great if @JeffBezanson would take a peek at #12544 and decide if it makes 0.4 or not.

But mostly I'm excited about getting it out the door!

**Comment by JeffBezanson** [comment]
Hopefully improving line numbers is not considered a breaking change, so not urgent for 0.4.0.

**Comment by timholy** [comment]
Sounds good to me.

**Comment by mlubin** [comment]
Given the [feature freeze](https://groups.google.com/forum/#!topic/julia-dev/M6XERBx96ts), time to bump to 0.4.0-pre?

**Comment by jskDr** [comment]
I am not sure whether the list is ordered by importance. However, I want to
say that ordered dict is highly important. As it is not supported Python
without using external library, dict() is convenient but complicated most
times.

James SJ Kim
sungjinkim@fas.harvard.edu

On Thu, Aug 27, 2015 at 8:20 PM, Miles Lubin <notifications@github.com>
wrote:

> Given the feature freeze
> <https://groups.google.com/forum/#!topic/julia-dev/M6XERBx96ts>, time to
> bump to 0.4.0-pre?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/issues/11536#issuecomment-135590122>.
>



-- 
Best regards,
(James) Sungjin Kim, Ph.D.
- Post-doc, CCB department in Harvard
  sungjinkim@fas.harvard.edu
  (Tech-consultant in Samsung Elec.)

**Comment by tkelman** [comment]
The list in the top post is meaningless at this point, features for 0.4 are frozen. Should probably close this issue since the scope has been determined, open a new one if necessary explicitly for tracking RC blockers.

**Comment by kmsquire** [comment]
@jskDr, I don't quite understand your comment about python, but there is an `OrderedDict` implementation in `DataStructures.jl`.

**Comment by jskDr** [comment]
Oh, really. I didn't know that. I will use OrderedDict whenever I need.
I thought it does not exist in Julia because ordered dicts (#10116
<https://github.com/JuliaLang/julia/pull/10116>) is listed above as a work
item.
Anyhow it is so appreciate as letting me know OrderedDict in
DataStructures.jl.
- https://github.com/JuliaLang/DataStructures.jl

James SJ Kim
sungjinkim@fas.harvard.edu


On Thu, Aug 27, 2015 at 10:45 PM, Kevin Squire <notifications@github.com>
wrote:

> @jskDr <https://github.com/jskDr>, I don't quite understand your comment
> about python, but there is an OrderedDict implementation in
> DataStructures.jl.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/issues/11536#issuecomment-135617345>.
>



-- 
Best regards,
(James) Sungjin Kim, Ph.D.
- Post-doc, CCB department in Harvard
  sungjinkim@fas.harvard.edu
  (Tech-consultant in Samsung Elec.)

**Comment by kmsquire** [comment]
(Mostly off-topic, sorry.)

The "ordered dict" item above was whether or not to replace the current `Dict` implementation in Base with an ordered dictionary implementation, which (if implemented correctly) can make traversal faster, but which has additional overhead in other ways.  This has been decided against (because of the overhead), but it might be useful to replace the `OrderedDict` in DataStructures.jl with the one that Jeff created for testing.

Also, just to be clear, in Julia `OrderedDict`s maintain their *insertion* order.  `DataStructures.jl` also has a `SortedDict` type, which maintains keys in *sorted* order.

(For anyone coming from C++, this might be confusing, as the default `map` class there is described as "ordered", meaning "sorted", and there is also an `unordered_map`.)
